internal/proto/encoding.go

```diff
@@ -0,0 +1,250 @@
+	ln := len(v)
+	if ln == 0 {
+		return 0
+	}
+	return protowire.SizeTag(num) + protowire.SizeBytes(ln)
+}
+
+// MarshalBytes encodes 'bytes' or 'string' protobuf field with given number and
+// value into b and returns the number of bytes written. If the buffer is too
+// small, MarshalBytes will panic.
+func MarshalBytes[T Bytes](b []byte, num protowire.Number, v T) int {
+	if len(v) == 0 {
+		return 0
+	}
+	off := binary.PutUvarint(b, protowire.EncodeTag(num, protowire.BytesType))
+	off += binary.PutUvarint(b[off:], uint64(len(v)))
+	return off + copy(b[off:], v)
+}
+
+// SizeFixed32 returns the encoded size of 'fixed32' protobuf field with given
+// number and value.
+func SizeFixed32(num protowire.Number, v uint32) int {
+	if v == 0 {
+		return 0
+	}
+	return protowire.SizeTag(num) + protowire.SizeFixed32()
+}
+
+// MarshalFixed32 encodes 'fixed32' protobuf field with given number and value
+// into b and returns the number of bytes written. If the buffer is too small,
+// MarshalFixed32 will panic.
+func MarshalFixed32(b []byte, num protowire.Number, v uint32) int {
+	if v == 0 {
+		return 0
+	}
+	off := binary.PutUvarint(b, protowire.EncodeTag(num, protowire.Fixed32Type))
+	binary.LittleEndian.PutUint32(b[off:], v)
+	return off + protowire.SizeFixed32()
+}
+
+// SizeFixed64 returns the encoded size of 'fixed64' protobuf field with given
+// number and value.
+func SizeFixed64(num protowire.Number, v uint64) int {
+	if v == 0 {
+		return 0
+	}
+	return protowire.SizeTag(num) + protowire.SizeFixed64()
+}
+
+// MarshalFixed64 encodes 'fixed64' protobuf field with given number and value
+// into b and returns the number of bytes written. If the buffer is too small,
+// MarshalFixed64 will panic.
+func MarshalFixed64(b []byte, num protowire.Number, v uint64) int {
+	if v == 0 {
+		return 0
+	}
+	off := binary.PutUvarint(b, protowire.EncodeTag(num, protowire.Fixed64Type))
+	binary.LittleEndian.PutUint64(b[off:], v)
+	return off + protowire.SizeFixed64()
+}
+
+// SizeFloat returns the encoded size of 'float' protobuf field with given
+// number and value.
+func SizeFloat(num protowire.Number, v float32) int {
+	return SizeFixed32(num, math.Float32bits(v))
+}
+
+// MarshalFloat encodes 'float' protobuf field with given number and value into
+// b and returns the number of bytes written. If the buffer is too small,
+// MarshalFloat will panic.
+func MarshalFloat(b []byte, num protowire.Number, v float32) int {
+	return MarshalFixed32(b, num, math.Float32bits(v))
+}
+
+// SizeDouble returns the encoded size of 'double' protobuf field with given
+// number and value.
+func SizeDouble(num protowire.Number, v float64) int {
+	return SizeFixed64(num, math.Float64bits(v))
+}
+
+// MarshalDouble encodes 'double' protobuf field with given number and value
+// into b and returns the number of bytes written. If the buffer is too small,
+// MarshalDouble will panic.
+func MarshalDouble(b []byte, num protowire.Number, v float64) int {
```

Comment:
why is it `Double`? api-go has it as `float64` as i remember. i expect this package to be a translator b/w "go language" and "profobuff language"

Reply:
these are protobuf types, so `double`

Reply:
they are marshallers of the go's types into some raw data, it is `MarshalDouble(... v float64)`, why? it was like [this](https://github.com/nspcc-dev/neofs-api-go/blob/a75c9e415c7681937b40ccc2297f6b691488a86a/util/proto/marshal.go#L357) before and i agree with it

Reply:
not _some raw data_ but protobuf. Any `bool`, `proto.Integer` or `float32` can be marshalled into `double`, and `float64` - into `uint64` or `fixed64` or even bytes. With this, `Double` wins

what i can suggest is to call it `MarshalToDouble` - would this reduce ur doubts?

Reply:
> Any bool, proto.Integer or float32 can be marshalled into double, and float64 - into uint64 or fixed64 or even bytes

do we marshal or plan to marshal `bool`, etc. into `double`? all i want is to have it as clear as possible. i expect a go struct that stores fields for `double` to have it `float64`. anyway, you do not call it `MarshalFloat64ToDouble`, so you do not plan to marshal anything except `float64` (me neither), why isn't it `MarshalFloat64` (a func that chooses the most appropriate container (`double`) for the provided go type)?

Reply:
following ur logic i cannot call `MarshalFixed32/64` or `MarshalVarint` since there are no such types in Go. And there cannot be `MarshalUint64` cuz it can be var- or fixed-len. To sum up, imo protobuf type must be present in the naming if we want any clarity. And i dont think `MarshalXToY` worth spending time

Reply:
idk, lets call @roman-khimov

Reply:
If we're converting X to Y then it's either `MarshalToY` or `MarshalX`, where X is rather obvious from the function signature while Y is not so much. At the same time a number of X can be marshaled into Y, so while `MarshalToY` is appealing it can be problematic longterm. `marshal.X` or `toproto.X` is an option too, btw.

> And there cannot be MarshalUint64 cuz it can be var- or fixed-len.

In practice you choose encoding for your project once and not care about it later.

Reply:
> In practice you choose encoding for your project once and not care about it later.

im talking about Go->protobuf `uint64`->`fixed64` and `uint64`->`uint64`. Two different Y making MarshalX defective

Reply:
> marshal.X or toproto.X is an option too, btw

packages? and where size funcs then?

Reply:
also guys note that this is internal package so we can rename funcs whenever we want

Reply:
`MarshalToY` is a good compromise for me that i can accept. but i have to mention that is not a flexible approach when we are talking about two different types (e.g `float64` and `float32` to `double`)

Reply:
example: [this](https://github.com/nspcc-dev/neofs-api-go/blob/a75c9e415c7681937b40ccc2297f6b691488a86a/audit/grpc/types.pb.go#L35) and [this](https://github.com/nspcc-dev/neofs-api-go/blob/a75c9e415c7681937b40ccc2297f6b691488a86a/session/grpc/types.pb.go#L431)

Reply:
renamed